### PR TITLE
Add log linear approx for GCY

### PR DESCRIPTION
**Pull request by Smit-create**
https://github.com/jstac/sdfs_via_autodiff/issues/9#issuecomment-1647901414

### CODE REVIEW

File: code/gcy/gcy_model.py [hunk 1]
```diff
@@ -58,16 +62,98 @@ def __init__(self,
...
+                + fk1(x) * fAzπ(x) * (φ_zπ)**2 * (1-ρ_zπ)\
+                + (θ/2) * ((fk1(x) * fAλ(x)+1)**2 * s_λ**2 \
+                + (fk1(x) * fAz(x) * s_wx)**2+(fk1(x) * fAc(x) * s_wc)**2\
+                + (fk1(x) * fAzπ(x) * s_wxπ)**2)) \
+                / (1-fk1(x))
+
+    def fq_bar(x):
+        return x - fA0(x) - fAc(x)*(φ_c)**2 - fAz(x)*(φ_z)**2 - fAzπ(x)*(φ_zπ)**2
+
+    qbar = brentq(fq_bar, -20, 20)
```
Comment (Smit-create) [inline code/gcy/gcy_model.py:136]: I tested this by plotting the graph and the root is: `6.061309329603693` at default GCY params

**Comment by JunnanZ** [comment]
@Smit-create Great work! I've yet to carefully check the details of your implementation, but Newton's method now works with the loglinear WC ratio as the initial guess. It's 3x faster than successive approximation, not great, but a big step forward.

**Comment by Smit-create** [comment]
That's great news! Thank you @JunnanZ

**Comment by jstac** [comment]
Great work @Smit-create  and very good intuition from @JunnanZ regarding the need for a better initial condition.

On top of that we can now compare the log-linear solution with the true global solution and emphasize any economically relevant differences
(CC @olee071 )

**Review by jstac — APPROVED** [review]

**Comment by Smit-create** [comment]
Thanks, @jstac for the approval. @JunnanZ Please feel free to merge this or let me know if any other changes are required. Many thanks in advance!

File: code/gcy/gcy_model.py [hunk 2]
```diff
@@ -58,16 +62,98 @@ def __init__(self,
...
+        return ρ_λ/(1-fk1(x)*ρ_λ)
+
+    def fAπ(x):
+        return fk1(x)*(1-1/ψ)*ρ_π/((1-fk1(x)*ρ)*(1-fk1(x)*ρ_ππ))
+
+    def fAz(x):
+        return (θ/2)*(fk1(x)*fA1(x))**2/(1-fk1(x)*ρ_z)
+
+    def fAzπ(x):
+        return (θ/2)*(fk1(x)*fAπ(x))**2/(1-fk1(x)*ρ_zπ)
```
[the remark this answers is not in the document]
Comment (JunnanZ) [reply code/gcy/gcy_model.py:118]: I derived the coefficients myself and he paper is wrong. Your code (and the SSY code) is correct. I will merge the PR.

**Review by JunnanZ — APPROVED** [review]

File: code/gcy/gcy_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Smit-create) [reply code/gcy/gcy_model.py:118]: Perfect, thank you!

**Pull request merged**